controllers/webhooks/networking/cfroute_validation.go

```diff
@@ -59,188 +58,146 @@ var logger = logf.Log.WithName("route-validation")
 
 //+kubebuilder:webhook:path=/validate-korifi-cloudfoundry-org-v1alpha1-cfroute,mutating=false,failurePolicy=fail,sideEffects=None,groups=korifi.cloudfoundry.org,resources=cfroutes,verbs=create;update;delete,versions=v1alpha1,name=vcfroute.korifi.cloudfoundry.org,admissionReviewVersions={v1,v1beta1}
 
-type CFRouteValidation struct {
-	decoder            *admission.Decoder
-	rootNamespace      string
+type CFRouteValidator struct {
 	duplicateValidator NameValidator
-	Client             client.Client
+	rootNamespace      string
+	client             client.Client
 }
 
-func NewCFRouteValidation(nameValidator NameValidator, rootNamespace string, client client.Client) *CFRouteValidation {
-	return &CFRouteValidation{
+var _ webhook.CustomValidator = &CFRouteValidator{}
+
+func NewCFRouteValidator(nameValidator NameValidator, rootNamespace string, client client.Client) *CFRouteValidator {
+	return &CFRouteValidator{
 		duplicateValidator: nameValidator,
 		rootNamespace:      rootNamespace,
-		Client:             client,
+		client:             client,
 	}
 }
 
-func (v *CFRouteValidation) SetupWebhookWithManager(mgr ctrl.Manager) error {
-	mgr.GetWebhookServer().Register("/validate-korifi-cloudfoundry-org-v1alpha1-cfroute", &webhook.Admission{Handler: v})
+func (v *CFRouteValidator) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&korifiv1alpha1.CFRoute{}).
+		WithValidator(v).
+		Complete()
+}
 
-	return nil
+func (v *CFRouteValidator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	route, ok := obj.(*korifiv1alpha1.CFRoute)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a CFRoute but got a %T", obj))
+	}
+
+	domain, err := v.validateRoute(ctx, route)
+	if err != nil {
+		return err
+	}
+
+	return generateDuplicateValidationError(route, domain, v.duplicateValidator.ValidateCreate(ctx, logger, v.rootNamespace, uniqueName(*route)))
```

Comment:
It might improve readability to extract the duplicateValidator call from the method arg here. Then I would also be inclined to do the nil check here (even though that would be repeated in the update and delete validation methods) and remove the nil check from the generateDuplicateValidationError method to make it more focused.

Reply:
I was contemplating moving the logic from `generateDuplicateValidationError` into the `duplicateValidator.Validate*` methods as there seemed to be a lot of duplication of handling of the error across all the webhooks.